crowbar_framework/app/models/api/node.rb

```diff
@@ -150,9 +150,13 @@ def save_node_state(role, state = "upgrading")
       )
       if state == "upgraded"
         progress = status.progress
-        remaining = progress[:remaining_nodes] - 1
-        upgraded = progress[:upgraded_nodes] + 1
-        ::Crowbar::UpgradeStatus.new.save_nodes(upgraded, remaining)
+        # This should not really happen, but in some corner cases,
+        # repeated upgrade of the node could have been invoked
+        if progress[:remaining_nodes] > 0
+          remaining = progress[:remaining_nodes] - 1
+          upgraded = progress[:upgraded_nodes] + 1
+          ::Crowbar::UpgradeStatus.new.save_nodes(upgraded, remaining)
+        end
```

Comment:
Yes and no. In fact that "corner case" was introduced by the same commit - `finalize_node_upgrade` removes "node_upgrade_state" which makes it impossible to check if node is `upgraded?` Sure, it should not be needed after `finalize_node_upgrade`, but..